central/tlsconfig/manager_impl.go

```diff
@@ -149,7 +149,7 @@ func (m *managerImpl) TLSConfigurer(opts Options) (verifier.TLSConfigurer, error
 	if opts.RequireClientCert {
 		rootCfg.ClientAuth = tls.RequireAndVerifyClientCert
 	}
-	configurer := certwatch.NewTLSConfigHolder(rootCfg)
+	configurer := certwatch.NewTLSConfigHolder(rootCfg, tls.NoClientCert)
```

Comment:
I think one thing that _may_ be confusing is that this here AFAICT isn't necessarily the TLS config for sensor connections, but rather the one based on [the endpoint config file](https://github.com/stackrox/stackrox/blob/master/image/templates/helm/stackrox-central/config/central/endpoints.yaml.default) contents and by default the client auth is set to `RequireAndVerifyClientCert` which makes sense due to mTLS for central <-> sensor as well as, if configured, user PKI auth provider.

I'd say the reason why this is causing trouble now is because you have a race-condition in the first call to `UpdateTLSConfig`: you call it twice, either on updates to the server certificates or on updates to the contents of the config map containing the CA. Assumption would be that reading the server certificates within the directories will be faster than obtaining the CA and its contents from Kubernetes, hence the first call to `UpdateTLSConfig` will be when the server certificates are received (and no client CAs are set), thus the `ClientAuth` option is unset / set to `NoClientCert`.

While the behavior isn't what one would expect, it's somewhat understandable why it has been like this before: the mount is either there, or it's not.

Looking at it back now, in your case we are essentially lazy loading the TLS config at the moment, and the initial serve from the metrics endpoint may or may not contain all required certificates (i.e. the `NewTLSConfigurerFromEnv` call). Ideally, we'd have a "run once" watch for changes to be sure we at least attempted _once_ to fetch required certificates before returning the config in this case.

Reply:
The start up is not the only case where the fallback matters. It also comes into play when the config map is deleted / changed such that no certs are found.

> Ideally, we'd have a "run once" watch for changes to be sure we at least attempted once to fetch required certificates before returning the config in this case.

What do we do then if the initial fetch fails? Tbh I don't see an issue with the lazy loading. The question is if incoming requests should be rejected (proposed behavior), or allowed (current behavior), if the certs are either not there or incomplete.

Reply:
> I think one thing that may be confusing is that this here AFAICT isn't necessarily the TLS config for sensor connections, but rather the one based on [the endpoint config file](https://github.com/stackrox/stackrox/blob/master/image/templates/helm/stackrox-central/config/central/endpoints.yaml.default) contents and by default the client auth is set to RequireAndVerifyClientCert which makes sense due to mTLS for central <-> sensor as well as, if configured, user PKI auth provider.

What I find unintuitive is that a TLS config holder is modifying static parts of the TLS config. I think it would make more sense to simply leave `ClientAuth` as it was set in the TLS root config. On the other hand changing this behavior now could cause unintended problems, so I would leave it as is.

Reply:
>What do we do then if the initial fetch fails? Tbh I don't see an issue with the lazy loading.

My point is that lazy loading here might be something we have to do due to the nature of reading asynchronously the CA from the config map, however I'd expect that we reject request until we can be sure that we've read the config map successfully _at least_ once and use the _expected_ TLS config. Right now, essentially you won't know which TLS config is being used after serving the first requests, and I don't think we should have this ambiguity here.

>What I find unintuitive is that a TLS config holder is modifying static parts of the TLS config. I think it would make more sense to simply leave ClientAuth as it was set in the TLS root config. On the other hand changing this behavior now could cause unintended problems, so I would leave it as is.

Yeah agreed that part definitely looks weird, but agreed we're better off to not change it for now.